File: mobile/TitanGo/lib/core/services/auth_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:get/get.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../constants/app_constants.dart';
import '../network/api_client.dart';
import '../network/api_checker.dart';
import '../../app.dart';

/// AuthService — manages the Titan Go worker session.
///
/// Login flow:
///   1. POST /api/fsm/v1/auth/login   (shared with FSMCore)
///   2. Store token + company_id + worker_id
///   3. Update ApiClient headers with tenant credentials
///   4. Register FCM token via PUT /api/nexus/v1/worker/fcm-token
///   5. Subscribe to FCM topics (company + worker)
class AuthService extends GetxService {
  final _secureStorage = const FlutterSecureStorage();
  late SharedPreferences _prefs;

  final _isLoggedIn = false.obs;
  bool get isLoggedIn => _isLoggedIn.value;

  int?    _workerId;
  int?    _companyId;
  String? _workerName;

  int?    get workerId   => _workerId;
  int?    get companyId  => _companyId;
  String? get workerName => _workerName;

  @override
  Future<AuthService> onInit() async {
    _prefs     = await SharedPreferences.getInstance();
    _workerId  = _prefs.getInt(AppConstants.keyWorkerId);
    _companyId = _prefs.getInt(AppConstants.keyCompanyId);
    _workerName = _prefs.getString(AppConstants.keyWorkerName);

    final token = await _secureStorage.read(key: AppConstants.keyToken);
    if (token != null && _companyId != null) {
      Get.find<ApiClient>().updateCredentials(
        token:     token,
        companyId: _companyId!,
      );
      _isLoggedIn.value = true;
    }
    return this;
  }

  // ──────────────────────────────────────────────────────────────────────────
  // Login
  // ──────────────────────────────────────────────────────────────────────────

  Future<bool> login(String email, String password) async {
    final res = await Get.find<ApiClient>().post(
      AppConstants.loginUri,
      body: {'email': email, 'password': password},
    );

    if (!res.isOk) {
      ApiChecker.handleError(res.body, res.statusCode);
      return false;
    }

    final data = res.body as Map<String, dynamic>;
    final token     = data['token'] as String;
    final workerMap = data['worker'] as Map<String, dynamic>;
    final workerId  = (workerMap['id'] as num).toInt();

    // Fetch worker profile to get company_id
    Get.find<ApiClient>().updateCredentials(token: token, companyId: 0);
    final profileRes = await Get.find<ApiClient>().get(AppConstants.workerMeUri);

    int companyId = 0;
    if (profileRes.isOk && profileRes.body is Map) {
      companyId = ((profileRes.body as Map)['company_id'] as num?)?.toInt() ?? 0;
    }

    await _persistSession(
      token:      token,
      workerId:   workerId,
      companyId:  companyId,
      workerName: workerMap['name']?.toString() ?? '',
    );

    _isLoggedIn.value = true;
    return true;
  }

  // ──────────────────────────────────────────────────────────────────────────
  // Logout
  // ──────────────────────────────────────────────────────────────────────────

  Future<void> logout() async {
    await Get.find<ApiClient>().post(AppConstants.logoutUri);
    clearSession();
    Get.offAllNamed(RouteNames.login);
  }

  void clearSession() {
    _secureStorage.delete(key: AppConstants.keyToken);
    _prefs.remove(AppConstants.keyWorkerId);
    _prefs.remove(AppConstants.keyCompanyId);
    _prefs.remove(AppConstants.keyWorkerName);
    Get.find<ApiClient>().clearCredentials();
    _isLoggedIn.value = false;
    _workerId  = null;
    _companyId = null;
  }

  // ──────────────────────────────────────────────────────────────────────────
  // Session persistence
  // ──────────────────────────────────────────────────────────────────────────

  Future<void> _persistSession({
    required String token,
    required int    workerId,
    required int    companyId,
    required String workerName,
  }) async {
    await _secureStorage.write(key: AppConstants.keyToken, value: token);
    await _prefs.setInt(AppConstants.keyWorkerId, workerId);
    await _prefs.setInt(AppConstants.keyCompanyId, companyId);
    await _prefs.setString(AppConstants.keyWorkerName, workerName);

    _workerId   = workerId;
    _companyId  = companyId;
    _workerName = workerName;

    Get.find<ApiClient>().updateCredentials(token: token, companyId: companyId);
  }
}

File: Modules/TitanGo/Resources/js/src/services/auth.service.ts
// Titan Go — AuthService
// Handles login/logout, session persistence, and company/device context.

import { AuthSession, Worker } from '../types';
import { apiClient } from './api.service';
import { storage } from '../utils/storage';
import { v4 as uuid } from '../utils/uuid';

const SESSION_KEY = 'titan_go_session';

interface LoginPayload {
  email: string;
  password: string;
}

interface LoginResponse {
  token: string;
  expires_at: string;
  worker: { id: number; name: string; email: string };
}

class AuthService {
  private session: AuthSession | null = null;

  constructor() {
    this.session = storage.get<AuthSession>(SESSION_KEY);
    if (this.session) {
      apiClient.setSession(this.session);
    }
  }

  getSession(): AuthSession | null {
    return this.session;
  }

  isAuthenticated(): boolean {
    if (!this.session) return false;
    // Check token expiry
    const exp = new Date(this.session.expiresAt).getTime();
    return Date.now() < exp;
  }

  async login(payload: LoginPayload): Promise<AuthSession> {
    const deviceId = storage.get<string>('titan_go_device_id') ?? uuid();
    storage.set('titan_go_device_id', deviceId);

    const res = await apiClient.post<LoginResponse>('/api/fsm/v1/auth/login', payload);

    const worker: Worker = {
      id:        res.worker.id,
      name:      res.worker.name,
      email:     res.worker.email,
      companyId: (res.worker as any).company_id ?? 0,
      deviceId,
    };

    const session: AuthSession = {
      token:     res.token,
      expiresAt: res.expires_at,
      worker,
      companyId: worker.companyId,
      deviceId,
    };

    this.session = session;
    storage.set(SESSION_KEY, session);
    apiClient.setSession(session);

    return session;
  }

  async logout(): Promise<void> {
    try {
      await apiClient.post('/api/fsm/v1/auth/logout');
    } catch {}
    this.session = null;
    storage.remove(SESSION_KEY);
    apiClient.setSession(null);
  }
}

export const authService = new AuthService();
